### PR TITLE
Relocate RPCErr to ripple/protocol - fix previous PR

### PR DESCRIPTION
**Pull request by legleux**
## High Level Overview of Change
Adds `STCurrency.h` to `libxrpl` for Clio
Fixes PR #4885 botched by force push

### Context of Change

Force push in updating PR #4885 undid intended changes.
Latest Clio build requires `STCurrency.h` in `libxrpl` package.

### API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [x] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

### CODE REVIEW

File: conanfile.py [hunk 1]
```diff
@@ -147,9 +147,9 @@ def package(self):
     def package_info(self):
         libxrpl = self.cpp_info.components['libxrpl']
         libxrpl.libs = [
-            'xrpl_core',
-            'ed25519',
-            'secp256k1',
+            'libxrpl_core.a',
+            'libed25519.a',
+            'libsecp256k1.a',
```
Comment (thejohnfreeman) [inline conanfile.py:152]: This is undoing some other [changes](https://github.com/XRPLF/rippled/pull/4831/files). Perhaps you need to merge `develop` into your branch first?

**Review by thejohnfreeman — CHANGES_REQUESTED** [review]

File: Builds/CMake/RippledCore.cmake [hunk 2]
```diff
@@ -251,94 +247,116 @@ install (
...
+    src/ripple/resource/impl/Charge.cpp
+    src/ripple/resource/impl/Consumer.cpp
+    src/ripple/resource/impl/Entry.h
+    src/ripple/resource/impl/Fees.cpp
+    src/ripple/resource/impl/Import.h
+    src/ripple/resource/impl/Key.h
+    src/ripple/resource/impl/Kind.h
+    src/ripple/resource/impl/Logic.h
+    src/ripple/resource/impl/ResourceManager.cpp
+    src/ripple/resource/impl/Tuning.h
```
Comment (thejohnfreeman) [inline Builds/CMake/RippledCore.cmake:332]: None of the `.cpp` files in this list should be installed. Please remove.

File: Builds/CMake/RippledCore.cmake [hunk 3]
```diff
@@ -251,94 +247,116 @@ install (
...
-    src/ripple/server/impl/io_list.h
-    src/ripple/server/impl/Door.h
-    src/ripple/server/impl/PlainHTTPPeer.h
-    src/ripple/server/impl/PlainWSPeer.h
     src/ripple/server/impl/BaseHTTPPeer.h
-    src/ripple/server/impl/BaseWSPeer.h
     src/ripple/server/impl/BasePeer.h
+    src/ripple/server/impl/BaseWSPeer.h
+    src/ripple/server/impl/Door.h
+    src/ripple/server/impl/JSONRPCUtil.cpp
```
Comment (thejohnfreeman) [inline Builds/CMake/RippledCore.cmake:350]: Another.

File: Builds/CMake/RippledCore.cmake [hunk 4]
```diff
@@ -251,94 +247,116 @@ install (
...
+    src/ripple/server/impl/JSONRPCUtil.h
     src/ripple/server/impl/LowestLayer.h
+    src/ripple/server/impl/PlainHTTPPeer.h
+    src/ripple/server/impl/PlainWSPeer.h
+    src/ripple/server/impl/Port.cpp
```
Comment (thejohnfreeman) [inline Builds/CMake/RippledCore.cmake:355]: Another.

**Review by thejohnfreeman — CHANGES_REQUESTED** [review]

**Review by thejohnfreeman — APPROVED** [review]

**Review by thejohnfreeman — CHANGES_REQUESTED** [review]
This PR undoes [this commit](https://github.com/XRPLF/rippled/commit/f9e365828a00900feca14483ad955d0ee6c49310). Please remove that change.

**Review by ximinez — CHANGES_REQUESTED** [review]
Ok, it took me way longer than expected, but I built the `install` target using `develop` and this branch, then compared the results:
```
$ diff -r install.develop/ install
Binary files install.develop/bin/rippled and install/bin/rippled differ
Only in install.develop/include/ripple/beast/unit_test: main.cpp
Only in install.develop/include/ripple: net
Only in install/include/ripple/protocol: RPCErr.h
Only in install/include/ripple/resource: Consumer.h
Only in install/include/ripple/resource: Disposition.h
Only in install/include/ripple/resource: Gossip.h
Only in install/include/ripple/resource: impl
Only in install/include/ripple/resource: ResourceManager.h
Only in install/include/ripple/resource: Types.h
Only in install/include/ripple/server/impl: JSONRPCUtil.h
diff -r install.develop/lib/cmake/ripple/RippleConfigVersion.cmake install/lib/cmake/ripple/RippleConfigVersion.cmake
12c12
< set(PACKAGE_VERSION "2.2.0-b1")
---
> set(PACKAGE_VERSION "2.2.0-b2")
18c18
<   if("2.2.0-b1" MATCHES "^([0-9]+)\\.")
---
>   if("2.2.0-b2" MATCHES "^([0-9]+)\\.")
24c24
<     set(CVF_VERSION_MAJOR "2.2.0-b1")
---
>     set(CVF_VERSION_MAJOR "2.2.0-b2")
Binary files install.develop/lib/libxrpl_core.a and install/lib/libxrpl_core.a differ
```
1. The differences in the binary files are expected, since I had to recompile.
2. I notice that `beast/unit_test/main.cpp` was removed from the `install` list. This may have been originally included intentionally because it is not built into `rippled`, so there's no other way to get it. Is anybody going to miss it?
3. The removal of `net` and addition of `protocol/RPCErr.h` are expected since that's the main point of this PR.
4. The addition of the other `.h` files in `resource` is probably good - they were probably overlooked before. However the newly added `impl` files aren't even exposed to the rest of `rippled`. Are we sure we want them exposed to external projects?
5. `server/impl/JSONRPCUtil.h` is already exposed in `rippled` to other modules, so this is probably fine.
6. Finally, the `RippleConfigVersion.cmake` changes are due the the version number change I've already complained about below.

It's probably beyond the scope of this PR, but we really should audit the exposed `impl` files to determine if they really need to be included in the `install` set, and if they do, move them *out* of the `impl` directory....

**Comment by legleux** [comment]
> 2\. beast/unit_test/main.cpp

@ximinez This is one thing I was unsure about whether it actually belonged or not.

I think @thejohnfreeman has the intention to export all these via globs so that none are forgotten about anymore but that would entail the audit Ed refers to.
Should I just close this PR or is that audit/export on too hazy of a timeline?

**Comment by ximinez** [comment]
> This is one thing I was unsure about whether it actually belonged or not. What would be the utility of providing this or not whether installed to a system or provided in a Conan package with `libxrpl`?

I can't think of much utility. IIRC, if one were to build and run `beast/unit_test/main.cpp`, it would only be able run the unit tests that are part of `beast`. Those are all already included in `rippled` AFAIK. It might be useful as sample / example code, but that can just as easily be obtained from Github.

> Should I just close this PR or is that audit/export on too hazy of a timeline?

I vote to keep this PR open. It's a smaller subset of changes that can be reviewed and merged quickly, and I believe unblocks something that @thejohnfreeman is working on.

**Review by ximinez — APPROVED** [review]
Looks good, and all the extraneous changes have been undone.

Here's what the `install` before and after looks like:
```
$ diff -r install.develop/ install
Binary files install.develop/bin/rippled.exe and install/bin/rippled.exe differ
Only in install.develop/include/ripple/beast/unit_test: main.cpp
Only in install.develop/include/ripple: net
Only in install/include/ripple/protocol: RPCErr.h
Only in install/include/ripple/resource: Consumer.h
Only in install/include/ripple/resource: Disposition.h
Only in install/include/ripple/resource: Gossip.h
Only in install/include/ripple/resource: ResourceManager.h
Only in install/include/ripple/resource: Types.h
Only in install/include/ripple/resource: impl
Only in install/include/ripple/server/impl: JSONRPCUtil.h
Binary files install.develop/lib/ed25519.lib and install/lib/ed25519.lib differ
Binary files install.develop/lib/secp256k1.lib and install/lib/secp256k1.lib differ
Binary files install.develop/lib/xrpl_core.lib and install/lib/xrpl_core.lib differ
```

File: Builds/CMake/RippledCore.cmake [hunk 5]
```diff
@@ -422,9 +435,9 @@ if (tests)
       src/ripple/beast/unit_test/reporter.hpp
       src/ripple/beast/unit_test/results.hpp
       src/ripple/beast/unit_test/runner.hpp
-      src/ripple/beast/unit_test/suite.hpp
       src/ripple/beast/unit_test/suite_info.hpp
       src/ripple/beast/unit_test/suite_list.hpp
+      src/ripple/beast/unit_test/suite.hpp
```
Comment (thejohnfreeman) [inline Builds/CMake/RippledCore.cmake:440]: The surrounding command, from line 427 (`if(tests)`) to line 442 (`DESTINATION ...`), should likely be removed. I don't know why we're installing to path `ripple/beast/extras/unit_test/`. Nothing should depend on that. We're already installing all of these headers to `ripple/beast/unit_test/` (lines 383-410).

File: Builds/CMake/RippledCore.cmake [hunk 6]
```diff
@@ -404,9 +416,10 @@ install (
   FILES
     src/ripple/beast/utility/Journal.h
     src/ripple/beast/utility/PropertyStream.h
+    src/ripple/beast/utility/WrappedSink.h
+    src/ripple/beast/utility/WrappedSink.h
```
Comment (thejohnfreeman) [inline Builds/CMake/RippledCore.cmake:420]: Doubled.

**Review by thejohnfreeman — CHANGES_REQUESTED** [review]

**Comment by thejohnfreeman** [comment]
@ximinez we're already installing some headers in `impl` directories. I would like to hold off on moving them around. My plan is to remove all `impl` directories in the PR that separates libxrpl and xrpld sources.

**Review by thejohnfreeman — APPROVED** [review]

**Review by thejohnfreeman — APPROVED** [review]

**Pull request merged**